modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/ResolverCache.scala

```diff
@@ -149,12 +197,22 @@ object ResolverCache {
     c: LruMap[F, K, (TTL, Lookup[A])],
     schemaKey: K,
     freshResult: Lookup[A]
-  ): F[Lookup[A]] =
+  ): F[Lookup[A]] = putItemResult(c, schemaKey, freshResult).map(_.map(_.value))
+
+  /**
+   * Much like `putItem` but carries information about whether the cache was updated as a
+   * consequence, and the timestamp associated with the cached item.
+   */
+  private def putItemResult[F[_]: Monad: Clock, A, K](
+    c: LruMap[F, K, (TTL, Lookup[A])],
+    schemaKey: K,
+    freshResult: Lookup[A]
+  ): F[Either[LookupFailureMap, TimestampedItem[A]]] =
     for {
-      currentTime <- Clock[F].realTime
-      cached      <- c.get(schemaKey) // Ignore TTL invalidation
-      result = chooseResult(cached.map(_._2), freshResult)
-      _ <- c.put(schemaKey, (currentTime, result)) // Overwrite or bump TTL
+      seconds <- Clock[F].realTime
+      cached  <- c.get(schemaKey) // Ignore TTL invalidation
+      result = chooseResult(cached.map(_._2), freshResult, seconds)
+      _ <- c.put(schemaKey, (seconds, result.map(_.value))) // Overwrite or bump TTL
```

Comment:
There are no guarantees that `c` did not change since the `get` on the previous line.

Reply:
Hey @voropaevp you are right about the race conditions.  Our problem though is we need iglu-scala-client to work with `Id` effects, not just with `IO` effects.  Which means we can't use things like `Ref` or `Semaphore` or anything that uses `Concurrent`.

In this case, any race condition is harmless, because the cache always ends up eventually in a consistent state.  But you have to convince yourself of this just by looking at the code, there is no other way.  It is not a satisfying answer.

One improvement here could be to extend our [LruMap](https://github.com/snowplow-incubator/scala-lru-map/blob/0.6.0/src/main/scala/com.snowplowanalytics.lrumap/LruMap.scala) trait to have a `update` method.  For the `Id` instances the implementation would still call `get` and `set` sequentially because that's all it can do.  But then we have a `Concurrent` instance that uses a `Semaphore`.

However I don't think that adds much value.  There is something else we could do which really would add value.

Currently, if 20 fibers call `Resolver[F].lookupSchema(...)` at the same time, then the resolver makes 20 concurrent http requests to Iglu Server (if the cache is empty).  I would love to find a way to wrap that in a Semaphore, so we only make 1 request.  Without letting the Semaphore become a performance bottleneck, of course.

Reply:
Thanks for highlighting this @voropaevp. I think wrapping requests to Iglu Server in a semaphore sounds interesting @istreeter and thanks for elaborating on the need to support `Id` instances.

Re: adding an update function to scala-lru-map, interestingly there is a [get](https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/latest/com.github.benmanes.caffeine/com/github/benmanes/caffeine/cache/Cache.html) function in Caffeine which is also [in Scaffeine](https://github.com/blemale/scaffeine/blob/main/src/main/scala/com/github/blemale/scaffeine/Cache.scala#L49) that seems to be designed for this pattern. 

Here's a Scastie that demos a race condition using this get-then-set pattern: 
https://scastie.scala-lang.org/HCK2VLiFSB6cKHs7CSvWMQ 
And another that uses that get function that takes an update function, showing that only one update is attempted: 
https://scastie.scala-lang.org/nAgtuOIQTpyziHpag8v4ig
(record of both scasties also in a gist [here](https://gist.github.com/lmath/17c2470b33ffeefce9832070a6174c4e))

But I'm not sure we'd want Caffeine to handle this for us, and I'm also not sure it fits into the design principles of scala-lru-map.